gquant/dataframe_flow/_node_flow.py

```diff
@@ -138,61 +153,155 @@ def get_pout(df_out):
-            if (kcol in icols) and kval != icols[kcol]:
+                err_msg = \
+                    'Task "{}" missing required column "{}" '\
+                    'from "{}".'.format(self.uid, kcol, src_task)
+                out_err = '{}\n{}'.format(info_msg, err_msg)
+                raise LookupError(out_err)
+
+            ival = icols[kcol]
+            if kval != ival:
                 # special case for 'date'
-                if (kval == 'date' and icols[kcol]
+                if (kval == 'date' and ival
                         in ('datetime64[ms]', 'date', 'datetime64[ns]')):
-                    # continue
                     return
                 else:
-                    print("error for node %s, "
-                          "type %s mismatch %s"
-                          % (self.uid, kval, icols[kcol]))
+                    err_msg = 'Task "{}" column "{}" expected type "{}" got '\
+                        'type "{}" instead.'.format(self.uid, kcol, kval, ival)
+                    out_err = '{}\n{}'.format(info_msg, err_msg)
+                    raise LookupError(out_err)
 
         inputs_cols = self.get_input_columns()
 
-        # to_port (iport usually used as variable) is always set. Refer to
-        # TaskGraph.build method. In non-port case inputs are enumerated
-        # in the order that inputs are listed in the task spec. The order
-        # idx is used as an ad-hoc ports that aren't used. Below the data
-        # structure of inputs_cols is flattened.
-        incoming_cols = {
-            col_name: col_type for icol_dict in inputs_cols.values()
-            for col_name, col_type in icol_dict.items()
-        }
-        inputs_cols = incoming_cols
-
-        if self.required:
-            required = self.required
-            pinputs = self._task_obj[TaskSpecSchema.inputs]
-            for iport in self._get_input_ports():
-                required_iport = {
-                    col_name: col_type for col_name, col_type in
-                    required.get(iport, {}).items()}
-
-                # required_tran = self.__translate_column(required_iport)
-                required_tran = required_iport
-                if iport in inputs_cols:
-                    incoming_cols = inputs_cols[iport]
-                    in_taskid = pinputs[iport]
-
-                    for kcol, kval in required_tran.items():
-                        validate_required(incoming_cols, kcol, kval,
-                                          in_taskid, iport)
+        if not self.required:
+            return
+
+        required = self.required
+        for iport in self._get_input_ports():
+            if iport not in required:
+                continue
+            required_iport = required[iport]
+
+            if iport not in inputs_cols:
+                # Is it possible that iport not connected? If so iport should
+                # not be in required. Should raise an exception here.
+                warn_msg = \
```

Comment:
Since the logic is inside the 'run' function. It makes sense to throw an exception as the required input is not connected.

Reply:
Yes, but I didn't think this was an egregious mistake to warrant raising an exception. I hope that raising a warning would be enough to inform the user to fix their node. As an exampled I fixed this in `gquant/plugin_nodes/transform/normalizationNode.py`. That normalization node ran fine before the fix, just produced this warning. A node shouldn't list something in required if it's not really required.

On the other hand maybe it's useful to have required for a port even if the port is not required so that in the UI you might see required columns tooltip. I actually had another flag to indicate whether a port was optional: `PortsSpecSchema.optional`. But it's not used consistently. The port being optional is handled without this flag.

Reply:
Okay. Let's leave it as is.